File: ros2_ws/src/python_workspace/python_workspace/extermination_node.py
import cv2
import sys
import os

import rclpy
from rclpy.time import Time
from rclpy.node import Node
from rclpy.executors import MultiThreadedExecutor
from cv_bridge import CvBridge

from sensor_msgs.msg import Image
from std_msgs.msg import Header, Int8, String
from custom_interfaces.msg import InferenceOutput

from .scripts.utils import ModelInference
sys.path.append(os.path.abspath('/home/user/ROS/python_wip'))
from display import display_annotated_image

class ExterminationNode(Node):
    def __init__(self):
        super().__init__('extermination_node')
        
        self.get_logger().info("Initializing Extermination Node")
        
        self.declare_parameter('use_display_node', True)
        self.declare_parameter('camera_side', 'left')
        self.use_display_node = self.get_parameter('use_display_node').get_parameter_value().bool_value
        self.camera_side = self.get_parameter('camera_side').get_parameter_value().string_value
        
        self.window = "Left Camera" if self.camera_side == "left" else "Right Camera" if self.use_display_node else None
        print(self.window)
        
        self.publishing_rate = 1.0
        self.lower_range = [78, 158, 124]
        self.upper_range = [60, 255, 255]
        self.minimum_area = 100
        self.minimum_confidence = 0.5
        self.boxes_present = 0
        self.model = ModelInference()
        self.bridge = CvBridge()
        self.boxes_msg = Int8()
        self.boxes_msg.data = 0
        
        self.inference_subscription = self.create_subscription(InferenceOutput, f'{self.camera_side}_inference_output', self.inference_callback, 10)
        self.box_publisher = self.create_publisher(Int8, f'{self.camera_side}_extermination_output', 10)
        self.timer = self.create_timer(self.publishing_rate, self.timer_callback)
        self.image_publisher = self.create_publisher(Image, 'annotated_image', 10)

    def inference_callback(self, msg):
        self.get_logger().info("Received Bounding Boxes")
        
        preprocessed_image = self.bridge.imgmsg_to_cv2(msg.preprocessed_image, desired_encoding='passthrough') # what is this needed for?
        raw_image = self.bridge.imgmsg_to_cv2(msg.raw_image, desired_encoding='passthrough')
        bounding_boxes = self.model.postprocess(msg.confidences.data,msg.bounding_boxes.data, raw_image,msg.velocity)
        # final_image = self.model.draw_boxes(raw_image,bounding_boxes,velocity=msg.velocity)
        labels = [f"{i}%" for i in msg.confidences.data]
        final_image = display_annotated_image(raw_image, bounding_boxes, labels)
        image_msg = self.bridge.cv2_to_imgmsg(final_image, encoding='bgr8')
        self.image_publisher.publish(image_msg)
        # if self.use_display_node:
        #     cv2.imshow("left window", final_image)
        #     cv2.waitKey(1)

        if len(bounding_boxes) > 0:
            self.boxes_present = 1
        else:
            self.boxes_present = 0
        
        self.boxes_msg = Int8()
        self.boxes_msg.data = self.boxes_present
    
    def timer_callback(self):
        self.box_publisher.publish(self.boxes_msg)
        self.get_logger().info("Published results to Proxy Node")

def main(args=None):
    rclpy.init(args=args)
    extermination_node = ExterminationNode()
    # executor = MultiThreadedExecutor(num_threads=1)
    # executor.add_node(extermination_node)
    # try:
    #     executor.spin()
    # except KeyboardInterrupt:
    #     print("Shutting down extermination node")
    # finally:
    #     executor.shutdown()
    #     extermination_node.destroy_node()
    #     rclpy.shutdown()
    try:
        rclpy.spin(extermination_node)
    except KeyboardInterrupt:
        extermination_node.get_logger().info("shutting down")
    finally:
        extermination_node.destroy_node()
        rclpy.shutdown()
        cv2.destroyAllWindows()

if __name__ == '__main__':
    main()

File: ros2_ws/src/cpp_package/src/extermination_node.cpp
#include <rclcpp/rclcpp.hpp>
#include <sensor_msgs/msg/image.hpp>
#include <std_msgs/msg/int8.hpp>
#include <custom_interfaces/msg/inference_output.hpp>
#include <cv_bridge/cv_bridge.h>
#include <opencv2/opencv.hpp>
#include "utils/model_inference.hpp"

// C++ Equivalent of cv_bridge:
//      Used CvBridge to convert ROS Image messages to OpenCV Mat.
// Custom Utility Integration:
//      ModelInference class should provide postprocess and draw_boxes functions similar to Python.
// Multi-threaded Executor:
//      Ensures responsiveness by managing callbacks efficiently.

class ExterminationNode : public rclcpp::Node {
public:
    ExterminationNode() : Node("extermination_node") {
        RCLCPP_INFO(this->get_logger(), "Initializing Extermination Node");

        // Declare and retrieve parameters
        this->declare_parameter<bool>("use_display_node", true);
        this->declare_parameter<std::string>("camera_side", "left");
        use_display_node_ = this->get_parameter("use_display_node").as_bool();
        camera_side_ = this->get_parameter("camera_side").as_string();

        window_ = (camera_side_ == "left") ? "Left Camera" : "Right Camera";
        if (!use_display_node_) {
            window_ = "";
        }

        publishing_rate_ = 1.0;
        lower_range_ = cv::Scalar(78, 158, 124);
        upper_range_ = cv::Scalar(60, 255, 255);
        minimum_area_ = 100;
        minimum_confidence_ = 0.5;
        boxes_present_ = 0;

        model_ = std::make_shared<ModelInference>();
        bridge_ = std::make_shared<cv_bridge::CvBridge>();

        boxes_msg_.data = 0;

        // Create subscription and publisher
        inference_subscription_ = this->create_subscription<custom_interfaces::msg::InferenceOutput>(
            camera_side_ + "_inference_output", 10, 
            std::bind(&ExterminationNode::inference_callback, this, std::placeholders::_1));

        box_publisher_ = this->create_publisher<std_msgs::msg::Int8>(
            camera_side_ + "_extermination_output", 10);

        timer_ = this->create_wall_timer(
            std::chrono::milliseconds(static_cast<int>(1000 / publishing_rate_)),
            std::bind(&ExterminationNode::timer_callback, this));
    }

private:
    void inference_callback(const custom_interfaces::msg::InferenceOutput::SharedPtr msg) {
        RCLCPP_INFO(this->get_logger(), "Received Bounding Boxes");

        // Convert ROS image messages to OpenCV
        cv::Mat preprocessed_image = bridge_->imgmsg_to_cv2(msg->preprocessed_image, "passthrough");
        cv::Mat raw_image = bridge_->imgmsg_to_cv2(msg->raw_image, "passthrough");

        // Perform post-processing and draw bounding boxes
        auto bounding_boxes = model_->postprocess(
            msg->confidences, msg->bounding_boxes, raw_image, msg->velocity);

        cv::Mat final_image = model_->draw_boxes(raw_image, bounding_boxes, msg->velocity);

        // Display the image if required
        if (use_display_node_) {
            cv::imshow(window_, final_image);
            cv::waitKey(10);
        }

        // Update presence of bounding boxes
        boxes_present_ = !bounding_boxes.empty() ? 1 : 0;
        boxes_msg_.data = boxes_present_;
    }

    void timer_callback() {
        box_publisher_->publish(boxes_msg_);
        RCLCPP_INFO(this->get_logger(), "Published results to Proxy Node");
    }

    // Parameters
    bool use_display_node_;
    std::string camera_side_;
    std::string window_;

    // Model and preprocessing
    double publishing_rate_;
    cv::Scalar lower_range_;
    cv::Scalar upper_range_;
    int minimum_area_;
    double minimum_confidence_;
    int boxes_present_;

    std_msgs::msg::Int8 boxes_msg_;

    // Utilities
    std::shared_ptr<ModelInference> model_;
    std::shared_ptr<cv_bridge::CvBridge> bridge_;

    // ROS 2 communication
    rclcpp::Subscription<custom_interfaces::msg::InferenceOutput>::SharedPtr inference_subscription_;
    rclcpp::Publisher<std_msgs::msg::Int8>::SharedPtr box_publisher_;
    rclcpp::TimerBase::SharedPtr timer_;
};

int main(int argc, char **argv) {
    rclcpp::init(argc, argv);
    auto node = std::make_shared<ExterminationNode>();

    rclcpp::executors::MultiThreadedExecutor executor;
    executor.add_node(node);

    try {
        executor.spin();
    } catch (const std::exception &e) {
        RCLCPP_ERROR(node->get_logger(), "Shutting down Extermination Node: %s", e.what());
    }

    rclcpp::shutdown();
    return 0;
}
